src/tasks/DockerRunTaskProvider.ts

```diff
@@ -23,62 +23,41 @@ export interface DockerRunTask extends Task {
     definition: DockerRunTaskDefinition;
 }
 
-export class DockerRunTaskProvider implements TaskProvider {
-    constructor(private readonly helpers: { [key in DockerPlatform]: TaskHelper }) { }
-
-    public provideTasks(token?: CancellationToken): ProviderResult<Task[]> {
-        return []; // Intentionally empty, so that resolveTask gets used
-    }
-
-    public resolveTask(task: DockerRunTask, token?: CancellationToken): ProviderResult<Task> {
-        return callWithTelemetryAndErrorHandling(
-            'docker-run-resolve',
-            async (actionContext: IActionContext) => {
-                const taskPlatform = getPlatform(task.definition);
-                actionContext.telemetry.properties.platform = taskPlatform;
-
-                return await this.resolveTaskInternal(
-                    {
-                        folder: task.scope as WorkspaceFolder,
-                        platform: taskPlatform,
-                        actionContext: actionContext,
-                        cancellationToken: token,
-                    },
-                    task
-                );
-            }
-        );
-    }
-
-    // TODO: Can we skip if a recently-started container exists?
-    private async resolveTaskInternal(context: DockerRunTaskContext, task: DockerRunTask): Promise<Task> {
+export class DockerRunTaskProvider extends DockerTaskProviderBase {
+    constructor(helpers: { [key in DockerPlatform]: TaskHelper }) { super('docker-run', helpers) }
 
+    // TODO: Skip if container is freshly started, but probably depends on language
+    protected async executeTaskInternal(context: DockerRunTaskContext, task: DockerRunTask): Promise<void> {
         const definition = cloneObject(task.definition);
         definition.dockerRun = definition.dockerRun || {};
 
-        if (!context.folder) {
-            throw new Error(`Unable to determine task scope to execute docker-run task '${task.name}'.`);
-        }
+        context.actionContext.telemetry.properties.containerOS = definition.dockerRun.os || 'Linux';
 
         context.buildDefinition = await getAssociatedDockerBuildTask(task);
 
         const helper = this.getHelper(context.platform);
 
-        if (helper) {
-            definition.dockerRun = await helper.resolveDockerRunOptions(context, definition);
+        if (helper && helper.preRun) {
+            await helper.preRun(context, definition);
+            throwIfCancellationRequested(context);
         }
 
-        await this.validateResolvedDefinition(context, definition.dockerRun);
+        if (helper) {
+            definition.dockerRun = await helper.getDockerRunOptions(context, definition);
+            await this.validateResolvedDefinition(context, definition.dockerRun);

```

Comment:
As above, is there use if validating the definition regardless of whether there's a helper?

Reply:
(Same as above, made it this way by mistake--need to change to what you're describing)